controllers/etcdrestore_controller.go

```diff
@@ -0,0 +1,524 @@
+	// All restore pods exist and our ours, so have they finished? This would be the join part of the "funky fork/join".
+	anyFailed := false
+	allSuccess := true
+	for _, restorePod := range restorePods {
+		phase := restorePod.Status.Phase
+		if phase == corev1.PodSucceeded {
+			// Success is what we want. Continue with reconciliation.
+			log.Info("Restore Pod has succeeded", "pod-name", name(restorePod))
+		} else if phase == corev1.PodFailed || phase == corev1.PodUnknown {
+			// Pod has failed, so we fail
+			log.Info("Restore Pod was not successful", "pod-name", name(restorePod), "phase", phase)
+			allSuccess = false
+			anyFailed = true
+		} else {
+			// This covers the "Pending" and "Running" phases. Do nothing and wait for the Pod to finish.
+			log.Info("Restore Pod still running.", "pod-name", name(restorePod), "phase", phase)
+			allSuccess = false
+		}
+	}
+
+	if anyFailed {
+		// TODO Report error reason
+		log.Info("One or more restore pods were not successful. Marking the restore as failed.")
+		restore.Status.Phase = etcdv1alpha1.EtcdRestorePhaseFailed
+		err := r.Client.Status().Update(ctx, &restore)
+		return ctrl.Result{}, err
+	}
+	if !allSuccess {
+		log.Info("One or more restore pods are still executing, and none have failed. Waiting...")
+		// Hrm. Not all finished yet. Exit and do nothing.
+		return ctrl.Result{}, nil
+	}
+
+	// Create the cluster
+	log.Info("All restore pods are complete, creating etcd cluster")
+	cluster := etcdv1alpha1.EtcdCluster{}
+	err := r.Get(ctx, name(expectedCluster), &cluster)
+	if apierrors.IsNotFound(err) {
+		// No Cluster. Create it
+		log.Info("Creating new cluster", "cluster-name", name(expectedCluster))
+		err := r.Create(ctx, expectedCluster)
+		return ctrl.Result{}, err
+	} else if err != nil {
+		log.Info("Error creating cluster", "cluster-name", name(expectedCluster))
+		return ctrl.Result{}, err
+	}
+
+	// This is the end. Our cluster exists.
+	log.Info("Verified cluster exists, exiting.")
+	restore.Status.Phase = etcdv1alpha1.EtcdRestorePhaseCompleted
+	err = r.Client.Status().Update(ctx, &restore)
+	return ctrl.Result{}, err
+}
+
+func restorePodName(peer etcdv1alpha1.EtcdPeer) string {
+	return fmt.Sprintf("%s-restore", peer.GetName())
+}
+
+func restorePodNamespacedName(peer etcdv1alpha1.EtcdPeer) types.NamespacedName {
+	return types.NamespacedName{
+		Namespace: peer.GetNamespace(),
+		Name:      restorePodName(peer),
+	}
+}
+
+func stripPortFromURL(advertiseURL *url.URL) string {
+	return strings.Split(advertiseURL.Host, ":")[0]
+}
+
+func podForRestore(restore etcdv1alpha1.EtcdRestore,
+	peer etcdv1alpha1.EtcdPeer,
+	pvcName string,
+	restoreImage string) *corev1.Pod {
+
+	pod := corev1.Pod{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:            restorePodName(peer),
+			Namespace:       restore.Namespace,
+			OwnerReferences: []metav1.OwnerReference{*metav1.NewControllerRef(&restore, etcdv1alpha1.GroupVersion.WithKind("EtcdRestore"))},
```

Comment:
Consider using controller-runtime `controllerutil.SetControllerReference` instead.

Reply:
Can we pull this out into a later issue? If we change how we set owner references it'd like to do it everywhere at once (e.g., the cluster controller creating `EtcdPeer`s, the peer controller creating `EtcdClusters`).

Because `SetControllerReference` can return an `error`, I'd also have to change all of the functions that create these resources to return an `error` too and implemented the associated handling code.